spreadsheets.js

```diff
@@ -0,0 +1,215 @@
+
+    // TODO: yes, this is oddly specific - but unless he changes his username ....
+    "steveharwell"
+];
+
+// A scheduling error (as opposed to other errors)
+function SchedulingError () {}
+SchedulingError.prototype = new Error();
+function PairingError () {}
+PairingError.prototype = new Error();
+
+// Make an attempt to parse a scheduling string. 
+//
+// first parameter is the string to parse
+// second parameter is an options dict specifying the round extrema options
+function parse_scheduling(string, options) {
+    // Do some basic preprocessing
+    string = string.replace(/[@\.,\(\)\<\>]/g, ' ');
+
+    var parts = string.split(" ");
+
+    // Filter it to remove all the noise
+    var filtered_parts = [];
+    parts.forEach(function(part) {
+        // Remove all lone 
+        if (WORDS_TO_IGNORE.indexOf(part.toLowerCase()) != -1) {
+            return;
+        }
+        filtered_parts.push(part);
+    });
+    if (filtered_parts.length < 3) {
+        console.log("Unable to parse scheduling string: " + string);
+        throw new SchedulingError();
+    }
+
+    // Now build up some possible strings and try a bunch of patterns
+    // to find a date in our range.
+    var extrema = get_round_extrema(options);
+    date_string = filtered_parts.slice(2).join(" ");
+    date_string = date_string.toLowerCase();
+    var date_strings = [];
+    date_strings.push(date_string)
+    var now = moment.utc();
+    year = now.year();
+    month = now.month()+1;
+    date_strings.push("" + year + "/" + date_string);
+    date_strings.push("" + year + "/" + month + "/" + date_string);
+
+    var dates = [];
+    date_strings.forEach(function(date_string) {
+        DATE_FORMATS.forEach(function(format) {
+            var date = moment.utc(date_string, format);
+            if (!date.isValid()) {
+                return;
+            }
+            if (date.isBefore(extrema[0]) || date.isAfter(extrema[1])) {
+                return;
+            }
+            dates.push(date);
+        });
+    });
+    if (dates.length < 1) {
+        console.log("Couldn't parse date: " + date_strings[0]);
+        throw new SchedulingError();
+    }
+
+    // strip out any punctuation from the usernames
+    var white = filtered_parts[0].replace(/[@\.,\(\):]/g, '');
+    var black = filtered_parts[1].replace(/[@\.,\(\):]/g, '');
+    // return the result
+    return {
+        white: white,
+        black: black,
+        date: dates[0]
+    };
+}
+
+// Calculate the extrema of the current round.
+function get_round_extrema(options) {
+    options = options || {};
+    if (!options.containing_date) {
+        round_start = moment.utc();
+    } else {
+        round_start = moment(options.containing_date).clone();
+    }
+
+    while (round_start.isoWeekday() != ISO_TUESDAY) {
+        round_start.subtract(1, 'days');
+    }
+    round_start.hour(0).minute(0).second(0);
+    round_end = round_start.clone().add(7, 'days');
+    if (options.offset_hours) {
```

Comment:
whats happening here? I dont understand what offset_hours is for? because we pair at a different time?